codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -168,16 +174,51 @@ public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf
                     frameHeader.release();
                     frameHeader = null;
                 }
-                ByteBuf frameHeader2 = ctx.alloc().buffer(FRAME_HEADER_LENGTH);
-                flags.endOfStream(endStream);
-                writeFrameHeaderInternal(frameHeader2, remainingData, DATA, flags, streamId);
-                ctx.write(frameHeader2, promiseAggregator.newPromise());
-
-                // Write the payload.
-                ByteBuf lastFrame = data.readSlice(remainingData);
-                data = null;
-                ctx.write(lastFrame, promiseAggregator.newPromise());
+                // if data is an heap buffer it should still requires Netty to copy it into a direct buffer
+                // before sending it to the socket, so let's perform this copy earlier
+                if (!data.hasArray() && remainingData > COPY_SLICE_CUTOFF_BYTES) {
```

Comment:
@bryce-anderson for heap buffers, differently from http 1.1 we don't risk to wrongly estimate the buffer, so it's safe to anticipate the copy